### PR TITLE
Directed graph layout and presentation

### PR DESCRIPTION
**Pull request by tiffon**
## High level

The overall effort for creating the directed graphs is being broken into multiple PRs and merged into an intermediary branch, `develop-directed-graph`, instead of `master`.

## "plexus"

>1. A network or interwoven mass, especially (anatomy) of nerves, blood vessels, or lymphatic vessels.
>
>2. (mathematics) The system of equations required for the complete expression of the relations which exist between a set of quantities.

[Wiktionary - plexus](https://en.wiktionary.org/wiki/plexus#Noun)

## Component

The directed graph will be used in:

- Service dependency graphs
- Viewing a trace as a DAG
- Comparing two traces via comparing two DAGs

This PR implements the functionality to generate the layout (positions and edges) of a graph, ~~but does not have the presentation functionality (React component), yet.~~ Also implemented in this PR is a React component to present the directed graph. It implements the MVP functionality that is required for kaiju.

Comments are very light and tests are TODO, but I wanted to get eyes on this as soon as possible.

### CODE REVIEW

File: packages/plexus/src/LayoutManager/Coordinator.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  idleWorkers: LayoutWorker[];
+  busyWorkers: LayoutWorker[];
+  callback: LayoutUpdate => void;
+
+  constructor(callback: LayoutUpdate => void) {
+    this.callback = callback;
+    this.currentLayout = null;
+    this.nextWorkerId = 1;
+    this.isDisposed = false;
+    this.idleWorkers = [this._initWorker()];
```
Comment (simonrobb) [inline packages/plexus/src/LayoutManager/Coordinator.js]: Why initialize with a worker? It looks like you just create one as they're required. Seems like introducing unnecessary state.

File: packages/plexus/src/LayoutManager/Coordinator.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  nextWorkerId: number;
+  isDisposed: boolean;
+  idleWorkers: LayoutWorker[];
+  busyWorkers: LayoutWorker[];
+  callback: LayoutUpdate => void;
+
+  constructor(callback: LayoutUpdate => void) {
+    this.callback = callback;
+    this.currentLayout = null;
+    this.nextWorkerId = 1;
```
Comment (simonrobb) [inline packages/plexus/src/LayoutManager/Coordinator.js]: A zero-based index might be more intuitive.

File: packages/plexus/src/LayoutManager/Coordinator.js [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    this.isDisposed = false;
+    this.idleWorkers = [this._initWorker()];
+    this.busyWorkers = [];
+  }
+
+  getLayout(id: number, edges: Edge[], vertices: SizeVertex[]) {
+    if (this.isDisposed) {
```
Comment (simonrobb) [inline packages/plexus/src/LayoutManager/Coordinator.js]: Any reason for having a disposed state? You're almost returning it to its initial state in the dispose method; it doesn't seem dangerous to allow a `getLayout` call on a disposed instance.

File: packages/plexus/src/LayoutManager/Coordinator.js (hunk 1 above)
Comment (tiffon) [reply packages/plexus/src/LayoutManager/Coordinator.js]: It takes some time to initialize the worker. A [quick test](https://fiddle.jshell.net/1ukjdxpb/show/light/) showed around 7 - 10ms. (Most of it is in generating the blob.) Kind of point-less to not do it lazily, though.

Will change.

As a side note: the rationale was driven by the fact that the workers are pooled.

File: packages/plexus/src/LayoutManager/Coordinator.js (hunk 3 above)
Comment (tiffon) [reply packages/plexus/src/LayoutManager/Coordinator.js]: Yeah, I guess the main thing is to interrupt any pending worker tasks, for instance if the view changes and the graph component unmounts.

Seems like having a `reset()` method to clear any current tasks would suffice. 

On the other hand, I don't know much about garbage collecting workers, so I'll leave a flag for killing idle workers, too.

**Comment by tiffon** [comment]
@simonrobb Thanks for the feedback, I made updates per your remarks (finally). Let me know if you see anything else that needs some work.

File: packages/plexus/src/LayoutManager/Coordinator.js (hunk 1 above)
Comment (simonrobb) [reply packages/plexus/src/LayoutManager/Coordinator.js]: This looks better to me 👍

File: packages/plexus/src/LayoutManager/Coordinator.js (hunk 2 above)
Comment (simonrobb) [reply packages/plexus/src/LayoutManager/Coordinator.js]: Were you intending to fix this?

**Pull request merged**